tables.go

```diff
@@ -241,15 +247,17 @@ func (l *State) next(t *table, key int) bool {
 		t.iterationKeys = keys
 	}
 	found := k == nil
-	for _, hk := range t.iterationKeys {
-		if found {
+	for i, hk := range t.iterationKeys {
+		if hk == nil { // skip deleted key
+		} else if _, present := t.hash[hk]; !present {
+			t.iterationKeys[i] = nil // mark key as deleted
```

Comment:
This removes references to deleted keys during iteration. It's not a perfect solution, since building a table, iterating over it, and nulling out all the entries will leak the keys as long as a reference exists to the table, if we never iterate over it again. It's a lot cheaper than sorting and searching, though, and much easier to implement (not least because we can't sort the keys in the general case).